File: jdk/src/share/classes/sun/nio/cs/UTF_8.java
/*
 * Copyright 2000-2006 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Sun designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Sun in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package sun.nio.cs;

import java.nio.ByteBuffer;
import java.nio.CharBuffer;
import java.nio.BufferOverflowException;
import java.nio.BufferUnderflowException;
import java.nio.charset.Charset;
import java.nio.charset.CharsetDecoder;
import java.nio.charset.CharsetEncoder;
import java.nio.charset.CoderResult;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.MalformedInputException;
import java.nio.charset.UnmappableCharacterException;


/*
 * # Bits   Bit pattern
 * 1    7   0xxxxxxx
 * 2   11   110xxxxx 10xxxxxx
 * 3   16   1110xxxx 10xxxxxx 10xxxxxx
 * 4   21   11110xxx 10xxxxxx 10xxxxxx 10xxxxxx
 * 5   26   111110xx 10xxxxxx 10xxxxxx 10xxxxxx 10xxxxxx
 * 6   31   1111110x 10xxxxxx 10xxxxxx 10xxxxxx 10xxxxxx 10xxxxxx
 *
 * UCS-2 uses 1-3, UTF-16 uses 1-4, UCS-4 uses 1-6
 */

class UTF_8 extends Unicode
{

    public UTF_8() {
        super("UTF-8", StandardCharsets.aliases_UTF_8);
    }

    public String historicalName() {
        return "UTF8";
    }

    public CharsetDecoder newDecoder() {
        return new Decoder(this);
    }

    public CharsetEncoder newEncoder() {
        return new Encoder(this);
    }


    private static class Decoder extends CharsetDecoder {
        private Decoder(Charset cs) {
            super(cs, 1.0f, 1.0f);
        }

        private boolean isContinuation(int b) {
            return ((b & 0xc0) == 0x80);
        }

        private final Surrogate.Generator sgg = new Surrogate.Generator();

        private CoderResult decodeArrayLoop(ByteBuffer src,
                                            CharBuffer dst)
        {
            byte[] sa = src.array();
            int sp = src.arrayOffset() + src.position();
            int sl = src.arrayOffset() + src.limit();
            assert (sp <= sl);
            sp = (sp <= sl ? sp : sl);
            char[] da = dst.array();
            int dp = dst.arrayOffset() + dst.position();
            int dl = dst.arrayOffset() + dst.limit();
            assert (dp <= dl);
            dp = (dp <= dl ? dp : dl);

            try {
                while (sp < sl) {
                    int b1 = sa[sp];
                    int b2, b3;
                    switch ((b1 >> 4) & 0x0f) {

                    case 0: case 1: case 2: case 3:
                    case 4: case 5: case 6: case 7:
                        // 1 byte, 7 bits: 0xxxxxxx
                        if (dl - dp < 1)
                            return CoderResult.OVERFLOW;
                        da[dp++] = (char)(b1 & 0x7f);
                        sp++;
                        continue;

                    case 12: case 13:
                        // 2 bytes, 11 bits: 110xxxxx 10xxxxxx
                        if (sl - sp < 2)
                            return CoderResult.UNDERFLOW;
                        if (dl - dp < 1)
                            return CoderResult.OVERFLOW;
                        if (!isContinuation(b2 = sa[sp + 1]))
                            return CoderResult.malformedForLength(1);
                        da[dp++] = ((char)(((b1 & 0x1f) << 6) |
                                           ((b2 & 0x3f) << 0)));
                        sp += 2;
                        continue;

                    case 14:
                        // 3 bytes, 16 bits: 1110xxxx 10xxxxxx 10xxxxxx
                        if (sl - sp < 3)
                            return CoderResult.UNDERFLOW;
                        if (dl - dp < 1)
                            return CoderResult.OVERFLOW;
                        if (!isContinuation(b2 = sa[sp + 1]))
                            return CoderResult.malformedForLength(1);
                        if (!isContinuation(b3 = sa[sp + 2]))
                            return CoderResult.malformedForLength(2);
                        da[dp++] = ((char)(((b1 & 0x0f) << 12) |
                                           ((b2 & 0x3f) << 06) |
                                           ((b3 & 0x3f) << 0)));
                        sp += 3;
                        continue;

                    case 15:
                        // 4, 5, or 6 bytes

                        int b4, b5, b6, uc, n;
                        switch (b1 & 0x0f) {

                        case 0: case 1: case 2: case 3:
                        case 4: case 5: case 6: case 7:
                            // 4 bytes, 21 bits
                            if (sl - sp < 4)
                                return CoderResult.UNDERFLOW;
                            if (!isContinuation(b2 = sa[sp + 1]))
                                return CoderResult.malformedForLength(1);
                            if (!isContinuation(b3 = sa[sp + 2]))
                                return CoderResult.malformedForLength(2);
                            if (!isContinuation(b4 = sa[sp + 3]))
                                return CoderResult.malformedForLength(3);
                            uc = (((b1 & 0x07) << 18) |
                                  ((b2 & 0x3f) << 12) |
                                  ((b3 & 0x3f) << 06) |
                                  ((b4 & 0x3f) << 00));
                            n = 4;
                            break;

                        case 8: case 9: case 10: case 11:
                            // 5 bytes, 26 bits
                            if (sl - sp < 5)
                                return CoderResult.UNDERFLOW;
                            if (!isContinuation(b2 = sa[sp + 1]))
                                return CoderResult.malformedForLength(1);
                            if (!isContinuation(b3 = sa[sp + 2]))
                                return CoderResult.malformedForLength(2);
                            if (!isContinuation(b4 = sa[sp + 3]))
                                return CoderResult.malformedForLength(3);
                            if (!isContinuation(b5 = sa[sp + 4]))
                                return CoderResult.malformedForLength(4);
                            uc = (((b1 & 0x03) << 24) |
                                  ((b2 & 0x3f) << 18) |
                                  ((b3 & 0x3f) << 12) |
                                  ((b4 & 0x3f) << 06) |
                                  ((b5 & 0x3f) << 00));
                            n = 5;
                            break;

                        case 12: case 13:
                            // 6 bytes, 31 bits
                            if (sl - sp < 6)
                                return CoderResult.UNDERFLOW;
                            if (!isContinuation(b2 = sa[sp + 1]))
                                return CoderResult.malformedForLength(1);
                            if (!isContinuation(b3 = sa[sp + 2]))
                                return CoderResult.malformedForLength(2);
                            if (!isContinuation(b4 = sa[sp + 3]))
                                return CoderResult.malformedForLength(3);
                            if (!isContinuation(b5 = sa[sp + 4]))
                                return CoderResult.malformedForLength(4);
                            if (!isContinuation(b6 = sa[sp + 5]))
                                return CoderResult.malformedForLength(5);
                            uc = (((b1 & 0x01) << 30) |
                                  ((b2 & 0x3f) << 24) |
                                  ((b3 & 0x3f) << 18) |
                                  ((b4 & 0x3f) << 12) |
                                  ((b5 & 0x3f) << 06) |
                                  ((b6 & 0x3f)));
                            n = 6;
                            break;

                        default:
                            return CoderResult.malformedForLength(1);

                        }

                        int gn = sgg.generate(uc, n, da, dp, dl);
                        if (gn < 0)
                            return sgg.error();
                        dp += gn;
                        sp += n;
                        continue;

                    default:
                        return CoderResult.malformedForLength(1);

                    }

                }

                return CoderResult.UNDERFLOW;
            } finally {
                src.position(sp - src.arrayOffset());
                dst.position(dp - dst.arrayOffset());
            }
        }

        private CoderResult decodeBufferLoop(ByteBuffer src,
                                             CharBuffer dst)
        {
            int mark = src.position();
            try {
                while (src.hasRemaining()) {
                    int b1 = src.get();
                    int b2, b3;
                    switch ((b1 >> 4) & 0x0f) {

                    case 0: case 1: case 2: case 3:
                    case 4: case 5: case 6: case 7:
                        // 1 byte, 7 bits: 0xxxxxxx
                        if (dst.remaining() < 1)
                            return CoderResult.OVERFLOW;
                        dst.put((char)b1);
                        mark++;
                        continue;

                    case 12: case 13:
                        // 2 bytes, 11 bits: 110xxxxx 10xxxxxx
                        if (src.remaining() < 1)
                            return CoderResult.UNDERFLOW;
                        if (dst.remaining() < 1)
                            return CoderResult.OVERFLOW;
                        if (!isContinuation(b2 = src.get()))
                            return CoderResult.malformedForLength(1);
                        dst.put((char)(((b1 & 0x1f) << 6) |
                                       ((b2 & 0x3f) << 0)));
                        mark += 2;
                        continue;

                    case 14:
                        // 3 bytes, 16 bits: 1110xxxx 10xxxxxx 10xxxxxx
                        if (src.remaining() < 2)
                            return CoderResult.UNDERFLOW;
                        if (dst.remaining() < 1)
                            return CoderResult.OVERFLOW;
                        if (!isContinuation(b2 = src.get()))
                            return CoderResult.malformedForLength(1);
                        if (!isContinuation(b3 = src.get()))
                            return CoderResult.malformedForLength(2);
                        dst.put((char)(((b1 & 0x0f) << 12) |
                                       ((b2 & 0x3f) << 06) |
                                       ((b3 & 0x3f) << 0)));
                        mark += 3;
                        continue;

                    case 15:
                        // 4, 5, or 6 bytes

                        int b4, b5, b6, uc, n;
                        switch (b1 & 0x0f) {

                        case 0: case 1: case 2: case 3:
                        case 4: case 5: case 6: case 7:
                            // 4 bytes, 21 bits
                            if (src.remaining() < 3)
                                return CoderResult.UNDERFLOW;
                            if (!isContinuation(b2 = src.get()))
                                return CoderResult.malformedForLength(1);
                            if (!isContinuation(b3 = src.get()))
                                return CoderResult.malformedForLength(2);
                            if (!isContinuation(b4 = src.get()))
                                return CoderResult.malformedForLength(3);
                            uc = (((b1 & 0x07) << 18) |
                                  ((b2 & 0x3f) << 12) |
                                  ((b3 & 0x3f) << 06) |
                                  ((b4 & 0x3f) << 00));
                            n = 4;
                            break;

                        case 8: case 9: case 10: case 11:
                            // 5 bytes, 26 bits
                            if (src.remaining() < 4)
                                return CoderResult.UNDERFLOW;
                            if (!isContinuation(b2 = src.get()))
                                return CoderResult.malformedForLength(1);
                            if (!isContinuation(b3 = src.get()))
                                return CoderResult.malformedForLength(2);
                            if (!isContinuation(b4 = src.get()))
                                return CoderResult.malformedForLength(3);
                            if (!isContinuation(b5 = src.get()))
                                return CoderResult.malformedForLength(4);
                            uc = (((b1 & 0x03) << 24) |
                                  ((b2 & 0x3f) << 18) |
                                  ((b3 & 0x3f) << 12) |
                                  ((b4 & 0x3f) << 06) |
                                  ((b5 & 0x3f) << 00));
                            n = 5;
                            break;

                        case 12: case 13:
                            // 6 bytes, 31 bits
                            if (src.remaining() < 5)
                                return CoderResult.UNDERFLOW;
                            if (!isContinuation(b2 = src.get()))
                                return CoderResult.malformedForLength(1);
                            if (!isContinuation(b3 = src.get()))
                                return CoderResult.malformedForLength(2);
                            if (!isContinuation(b4 = src.get()))
                                return CoderResult.malformedForLength(3);
                            if (!isContinuation(b5 = src.get()))
                                return CoderResult.malformedForLength(4);
                            if (!isContinuation(b6 = src.get()))
                                return CoderResult.malformedForLength(5);
                            uc = (((b1 & 0x01) << 30) |
                                  ((b2 & 0x3f) << 24) |
                                  ((b3 & 0x3f) << 18) |
                                  ((b4 & 0x3f) << 12) |
                                  ((b5 & 0x3f) << 06) |
                                  ((b6 & 0x3f)));
                            n = 6;
                            break;

                        default:
                            return CoderResult.malformedForLength(1);

                        }

                        if (sgg.generate(uc, n, dst) < 0)
                            return sgg.error();
                        mark += n;
                        continue;

                    default:
                        return CoderResult.malformedForLength(1);

                    }

                }
                return CoderResult.UNDERFLOW;
            } finally {
                src.position(mark);
            }
        }

        protected CoderResult decodeLoop(ByteBuffer src,
                                         CharBuffer dst)
        {
            if (src.hasArray() && dst.hasArray())
                return decodeArrayLoop(src, dst);
            else
                return decodeBufferLoop(src, dst);
        }

    }


    private static class Encoder extends CharsetEncoder {

        private Encoder(Charset cs) {
            super(cs, 1.1f, 4.0f);
        }

        public boolean canEncode(char c) {
            return !Surrogate.is(c);
        }

        private final Surrogate.Parser sgp = new Surrogate.Parser();

        private CoderResult encodeArrayLoop(CharBuffer src,
                                            ByteBuffer dst)
        {
            char[] sa = src.array();
            int sp = src.arrayOffset() + src.position();
            int sl = src.arrayOffset() + src.limit();
            assert (sp <= sl);
            sp = (sp <= sl ? sp : sl);
            byte[] da = dst.array();
            int dp = dst.arrayOffset() + dst.position();
            int dl = dst.arrayOffset() + dst.limit();
            assert (dp <= dl);
            dp = (dp <= dl ? dp : dl);

            try {
                while (sp < sl) {
                    char c = sa[sp];

                    if (c < 0x80) {
                        // Have at most seven bits
                        if (dp >= dl)
                            return CoderResult.OVERFLOW;
                        da[dp++] = (byte)c;
                        sp++;
                        continue;
                    }

                    if (!Surrogate.is(c)) {
                        // 2 bytes, 11 bits
                        if (c < 0x800) {
                            if (dl - dp < 2)
                                return CoderResult.OVERFLOW;
                            da[dp++] = (byte)(0xc0 | ((c >> 06)));
                            da[dp++] = (byte)(0x80 | ((c >> 00) & 0x3f));
                            sp++;
                            continue;
                        }
                        if (c <= '\uFFFF') {
                            // 3 bytes, 16 bits
                            if (dl - dp < 3)
                                return CoderResult.OVERFLOW;
                            da[dp++] = (byte)(0xe0 | ((c >> 12)));
                            da[dp++] = (byte)(0x80 | ((c >> 06) & 0x3f));
                            da[dp++] = (byte)(0x80 | ((c >> 00) & 0x3f));
                            sp++;
                            continue;
                        }
                    }

                    // Have a surrogate pair
                    int uc = sgp.parse(c, sa, sp, sl);
                    if (uc < 0)
                        return sgp.error();
                    if (uc < 0x200000) {
                        if (dl - dp < 4)
                            return CoderResult.OVERFLOW;
                        da[dp++] = (byte)(0xf0 | ((uc >> 18)));
                        da[dp++] = (byte)(0x80 | ((uc >> 12) & 0x3f));
                        da[dp++] = (byte)(0x80 | ((uc >> 06) & 0x3f));
                        da[dp++] = (byte)(0x80 | ((uc >> 00) & 0x3f));
                        sp += sgp.increment();
                        continue;
                    }
                    assert false;

                }
                return CoderResult.UNDERFLOW;
            } finally {
                src.position(sp - src.arrayOffset());
                dst.position(dp - dst.arrayOffset());
            }
        }

        private CoderResult encodeBufferLoop(CharBuffer src,
                                             ByteBuffer dst)
        {
            int mark = src.position();
            try {
                while (src.hasRemaining()) {
                    char c = src.get();

                    if (c < 0x80) {
                        // Have at most seven bits
                        if (!dst.hasRemaining())
                            return CoderResult.OVERFLOW;
                        dst.put((byte)c);
                        mark++;
                        continue;
                    }

                    if (!Surrogate.is(c)) {
                        if (c < 0x800) {
                            // 2 bytes, 11 bits
                            if (dst.remaining() < 2)
                                return CoderResult.OVERFLOW;
                            dst.put((byte)(0xc0 | ((c >> 06))));
                            dst.put((byte)(0x80 | ((c >> 00) & 0x3f)));
                            mark++;
                            continue;
                        }
                        if (c <= '\uFFFF') {
                            // 3 bytes, 16 bits
                            if (dst.remaining() < 3)
                                return CoderResult.OVERFLOW;
                            dst.put((byte)(0xe0 | ((c >> 12))));
                            dst.put((byte)(0x80 | ((c >> 06) & 0x3f)));
                            dst.put((byte)(0x80 | ((c >> 00) & 0x3f)));
                            mark++;
                            continue;
                        }
                    }

                    // Have a surrogate pair
                    int uc = sgp.parse(c, src);
                    if (uc < 0)
                        return sgp.error();
                    if (uc < 0x200000) {
                        if (dst.remaining() < 4)
                            return CoderResult.OVERFLOW;
                        dst.put((byte)(0xf0 | ((uc >> 18))));
                        dst.put((byte)(0x80 | ((uc >> 12) & 0x3f)));
                        dst.put((byte)(0x80 | ((uc >> 06) & 0x3f)));
                        dst.put((byte)(0x80 | ((uc >> 00) & 0x3f)));
                        mark += sgp.increment();
                        continue;
                    }
                    assert false;

                }
                return CoderResult.UNDERFLOW;
            } finally {
                src.position(mark);
            }
        }

        protected final CoderResult encodeLoop(CharBuffer src,
                                               ByteBuffer dst)
        {
            if (src.hasArray() && dst.hasArray())
                return encodeArrayLoop(src, dst);
            else
                return encodeBufferLoop(src, dst);
        }

    }

}

File: hotspot/src/share/vm/utilities/utf8.cpp
/*
 * Copyright (c) 1997, 2010, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 *
 */

#include "precompiled.hpp"
#include "utilities/utf8.hpp"

// Assume the utf8 string is in legal form and has been
// checked in the class file parser/format checker.
char* UTF8::next(const char* str, jchar* value) {
  unsigned const char *ptr = (const unsigned char *)str;
  unsigned char ch, ch2, ch3;
  int length = -1;              /* bad length */
  jchar result;
  switch ((ch = ptr[0]) >> 4) {
    default:
    result = ch;
    length = 1;
    break;

  case 0x8: case 0x9: case 0xA: case 0xB: case 0xF:
    /* Shouldn't happen. */
    break;

  case 0xC: case 0xD:
    /* 110xxxxx  10xxxxxx */
    if (((ch2 = ptr[1]) & 0xC0) == 0x80) {
      unsigned char high_five = ch & 0x1F;
      unsigned char low_six = ch2 & 0x3F;
      result = (high_five << 6) + low_six;
      length = 2;
      break;
    }
    break;

  case 0xE:
    /* 1110xxxx 10xxxxxx 10xxxxxx */
    if (((ch2 = ptr[1]) & 0xC0) == 0x80) {
      if (((ch3 = ptr[2]) & 0xC0) == 0x80) {
        unsigned char high_four = ch & 0x0f;
        unsigned char mid_six = ch2 & 0x3f;
        unsigned char low_six = ch3 & 0x3f;
        result = (((high_four << 6) + mid_six) << 6) + low_six;
        length = 3;
      }
    }
    break;
  } /* end of switch */

  if (length <= 0) {
    *value = ptr[0];    /* default bad result; */
    return (char*)(ptr + 1); // make progress somehow
  }

  *value = result;

  // The assert is correct but the .class file is wrong
  // assert(UNICODE::utf8_size(result) == length, "checking reverse computation");
  return (char *)(ptr + length);
}

char* UTF8::next_character(const char* str, jint* value) {
  unsigned const char *ptr = (const unsigned char *)str;
  /* See if it's legal supplementary character:
     11101101 1010xxxx 10xxxxxx 11101101 1011xxxx 10xxxxxx */
  if (is_supplementary_character(ptr)) {
    *value = get_supplementary_character(ptr);
    return (char *)(ptr + 6);
  }
  jchar result;
  char* next_ch = next(str, &result);
  *value = result;
  return next_ch;
}

// Count bytes of the form 10xxxxxx and deduct this count
// from the total byte count.  The utf8 string must be in
// legal form which has been verified in the format checker.
int UTF8::unicode_length(const char* str, int len) {
  int num_chars = len;
  for (int i = 0; i < len; i++) {
    if ((str[i] & 0xC0) == 0x80) {
      --num_chars;
    }
  }
  return num_chars;
}

// Count bytes of the utf8 string except those in form
// 10xxxxxx which only appear in multibyte characters.
// The utf8 string must be in legal form and has been
// verified in the format checker.
int UTF8::unicode_length(const char* str) {
  int num_chars = 0;
  for (const char* p = str; *p; p++) {
    if (((*p) & 0xC0) != 0x80) {
      num_chars++;
    }
  }
  return num_chars;
}

// Writes a jchar a utf8 and returns the end
static u_char* utf8_write(u_char* base, jchar ch) {
  if ((ch != 0) && (ch <=0x7f)) {
    base[0] = (u_char) ch;
    return base + 1;
  }

  if (ch <= 0x7FF) {
    /* 11 bits or less. */
    unsigned char high_five = ch >> 6;
    unsigned char low_six = ch & 0x3F;
    base[0] = high_five | 0xC0; /* 110xxxxx */
    base[1] = low_six | 0x80;   /* 10xxxxxx */
    return base + 2;
  }
  /* possibly full 16 bits. */
  char high_four = ch >> 12;
  char mid_six = (ch >> 6) & 0x3F;
  char low_six = ch & 0x3f;
  base[0] = high_four | 0xE0; /* 1110xxxx */
  base[1] = mid_six | 0x80;   /* 10xxxxxx */
  base[2] = low_six | 0x80;   /* 10xxxxxx */
  return base + 3;
}

void UTF8::convert_to_unicode(const char* utf8_str, jchar* unicode_str, int unicode_length) {
  unsigned char ch;
  const char *ptr = (const char *)utf8_str;
  int index = 0;

  /* ASCII case loop optimization */
  for (; index < unicode_length; index++) {
    if((ch = ptr[0]) > 0x7F) { break; }
    unicode_str[index] = ch;
    ptr = (const char *)(ptr + 1);
  }

  for (; index < unicode_length; index++) {
    ptr = UTF8::next(ptr, &unicode_str[index]);
  }
}

// Returns NULL if 'c' it not found. This only works as long
// as 'c' is an ASCII character
jbyte* UTF8::strrchr(jbyte* base, int length, jbyte c) {
  assert(length >= 0, "sanity check");
  assert(c >= 0, "does not work for non-ASCII characters");
  // Skip backwards in string until 'c' is found or end is reached
  while(--length >= 0 && base[length] != c);
  return (length < 0) ? NULL : &base[length];
}

bool UTF8::equal(jbyte* base1, int length1, jbyte* base2, int length2) {
  // Length must be the same
  if (length1 != length2) return false;
  for (int i = 0; i < length1; i++) {
    if (base1[i] != base2[i]) return false;
  }
  return true;
}

bool UTF8::is_supplementary_character(const unsigned char* str) {
  return ((str[0] & 0xFF) == 0xED) && ((str[1] & 0xF0) == 0xA0) && ((str[2] & 0xC0) == 0x80)
      && ((str[3] & 0xFF) == 0xED) && ((str[4] & 0xF0) == 0xB0) && ((str[5] & 0xC0) == 0x80);
}

jint UTF8::get_supplementary_character(const unsigned char* str) {
  return 0x10000 + ((str[1] & 0x0f) << 16) + ((str[2] & 0x3f) << 10)
                 + ((str[4] & 0x0f) << 6)  + (str[5] & 0x3f);
}


//-------------------------------------------------------------------------------------


int UNICODE::utf8_size(jchar c) {
  if ((0x0001 <= c) && (c <= 0x007F)) return 1;
  if (c <= 0x07FF) return 2;
  return 3;
}

int UNICODE::utf8_length(jchar* base, int length) {
  int result = 0;
  for (int index = 0; index < length; index++) {
    jchar c = base[index];
    if ((0x0001 <= c) && (c <= 0x007F)) result += 1;
    else if (c <= 0x07FF) result += 2;
    else result += 3;
  }
  return result;
}

char* UNICODE::as_utf8(jchar* base, int length) {
  int utf8_len = utf8_length(base, length);
  u_char* result = NEW_RESOURCE_ARRAY(u_char, utf8_len + 1);
  u_char* p = result;
  for (int index = 0; index < length; index++) {
    p = utf8_write(p, base[index]);
  }
  *p = '\0';
  assert(p == &result[utf8_len], "length prediction must be correct");
  return (char*) result;
}

char* UNICODE::as_utf8(jchar* base, int length, char* buf, int buflen) {
  u_char* p = (u_char*)buf;
  u_char* end = (u_char*)buf + buflen;
  for (int index = 0; index < length; index++) {
    jchar c = base[index];
    if (p + utf8_size(c) >= end) break;      // string is truncated
    p = utf8_write(p, base[index]);
  }
  *p = '\0';
  return buf;
}

void UNICODE::convert_to_utf8(const jchar* base, int length, char* utf8_buffer) {
  for(int index = 0; index < length; index++) {
    utf8_buffer = (char*)utf8_write((u_char*)utf8_buffer, base[index]);
  }
  *utf8_buffer = '\0';
}
